### PR TITLE
[blocked by #789] Fix/dataset check [wip]

### LINKED ISSUE #840: Handle abstract loader that doesn't have a dataset member
# Feature / Bug (?)

I have an abstract loader that chains multiple pytorch loaders when training on video sequences. Each small loader contains one sequence and the chained loader just use itertools.chain to chain them together. 

I cannot put all data in a single loader because it does not make sense to read images from two different sequences.
I cannot use an IterableDataset because that would still have the same problem.

https://github.com/PyTorchLightning/pytorch-lightning/blob/06242c200a318a37d1f882c786e60354ec04533f/pytorch_lightning/trainer/data_loading.py#L58

This assumes that a loader must have a dataset field, which I felt is too restrictive. I suggest putting a check to see if the loader has a dataset member before doing this check. My workaround, for now, is just to declare a dataset member in my ChainedLoader to be None.

# Fix

Since this kind of abstract loader does not have an explicit handle to a dataset, it should belong to the concept of an IterableDataset, which the user specifies the number of batches rather than a percentage. So a simple fix would be to put a check of `hasattr` somewhere in the same line.

### PR DESCRIPTION
**Pull request by ethanwharris**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CHANGELOG.md)?

## What does this PR do?
Fixes #840 - also touches part of #698 

- Changed Exception order so that MisconfigurationException will be called and given the correct error when using IterableDataset
- Removed dependency on `Dataloader.dataset` following #840 
- Changed TQDM call to prevent error when `num_training_batches == float('inf')`
- `num_training_batches =float('inf')` is now the default when train dataloader doesn't have `__len__` (in addition to when using an `IterableDataset`)
- Added test for training using an iterable

- Note that these only apply to the train dataloader, test / val using `IterableDataset` will still cause an error, but thought we should merge this first as stops a few annoying behaviours

## PR review
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/data_loading.py [hunk 1]
```diff
@@ -55,13 +55,24 @@ def init_train_dataloader(self, model):
         self.get_train_dataloader = model.train_dataloader
 
         # determine number of training batches
-        if EXIST_ITER_DATASET and isinstance(self.get_train_dataloader().dataset, IterableDataset):
+        self.is_iterable_train_dataloader = (
+            EXIST_ITER_DATASET and hasattr(self.get_train_dataloader(), 'dataset') and
```
Comment (Borda) [inline pytorch_lightning/trainer/data_loading.py]: some changes seem similar to #789...

File: pytorch_lightning/trainer/data_loading.py [hunk 2]
```diff
@@ -55,13 +55,24 @@ def init_train_dataloader(self, model):
...
 
-            self.num_training_batches = len(self.get_train_dataloader())
-            self.num_training_batches = int(self.num_training_batches * self.train_percent_check)
+            try:
+                self.num_training_batches = len(self.get_train_dataloader())
+                self.num_training_batches = int(self.num_training_batches *
+                                                self.train_percent_check)
+            except TypeError as e:
+                # Assume infinite data loading if the dataloader doesn't implement __len__
+                if 'has no len' in str(e):
```
Comment (Borda) [inline pytorch_lightning/trainer/data_loading.py]: we shall have test for this...

File: pytorch_lightning/trainer/data_loading.py (hunk 1 above)
Comment (ethanwharris) [reply pytorch_lightning/trainer/data_loading.py]: Yeah, hadn't seen that. Feel free to merge #789 through and I can rebase after :)

**Comment by ethanwharris** [comment]
Now handled in #955

**Pull request closed without merging**